zcash_extensions/src/transparent/demo.rs

```diff
@@ -0,0 +1,786 @@
+        // Call through to the generic builder.
+        self.txn_builder
+            .add_tze_output(self.extension_id, value, &Precondition::open(hash_1))
+            .map_err(DemoBuildError::BaseBuilderError)
+    }
+
+    /// Add a witness to a previous channel-opening precondition and a new channel-closing
+    /// precondition to the transaction under construction.
+    pub fn demo_transfer_to_close(
+        &mut self,
+        prevout: (OutPoint, TzeOut),
+        transfer_amount: Amount,
+        preimage_1: [u8; 32],
+        hash_2: [u8; 32],
+    ) -> Result<(), DemoBuildError<B::BuildError>> {
+        let h1 = hash_1(&preimage_1, &hash_2);
+
+        // eagerly validate the relationship between prevout.1 and preimage_1
+        match Precondition::from_payload(
+            prevout.1.precondition.mode,
+            &prevout.1.precondition.payload,
+        ) {
+            Ok(Precondition::Open(hash)) => {
+                if hash.0 != h1 {
+                    Err(DemoBuildError::TransferMismatch {
+                        expected: hash.0,
+                        actual: h1,
+                    })?
+                }
+            }
+            Ok(Precondition::Close(_)) => Err(DemoBuildError::ExpectedOpen)?,
+            Err(parse_failure) => Err(DemoBuildError::PrevoutParseFailure(parse_failure))?,
+        }
+
+        self.txn_builder
+            .add_tze_input(self.extension_id, open::MODE, prevout, move |_| {
+                Ok(Witness::open(preimage_1))
+            })
+            .map_err(DemoBuildError::BaseBuilderError)?;
+
+        self.txn_builder
+            .add_tze_output(
+                self.extension_id,
+                transfer_amount,
+                &Precondition::close(hash_2),
+            )
+            .map_err(DemoBuildError::BaseBuilderError)
+    }
+
+    /// Add a channel-closing witness to the transaction under construction.
+    pub fn demo_close(
+        &mut self,
+        prevout: (OutPoint, TzeOut),
+        preimage_2: [u8; 32],
+    ) -> Result<(), DemoBuildError<B::BuildError>> {
+        let hash_2 = {
+            let mut hash = [0; 32];
+            hash.copy_from_slice(Params::new().hash_length(32).hash(&preimage_2).as_bytes());
+            hash
+        };
+
+        // eagerly validate the relationship between prevout.1 and preimage_2
+        match Precondition::from_payload(
+            prevout.1.precondition.mode,
+            &prevout.1.precondition.payload,
+        ) {
+            Ok(Precondition::Open(_)) => Err(DemoBuildError::ExpectedClose)?,
+            Ok(Precondition::Close(hash)) => {
+                if hash.0 != hash_2 {
+                    Err(DemoBuildError::CloseMismatch {
+                        expected: hash.0,
+                        actual: hash_2,
+                    })?
+                }
+            }
+            Err(parse_failure) => Err(DemoBuildError::PrevoutParseFailure(parse_failure))?,
```

Comment:
This can just be `return Err(DemoBuildError::PrevoutParseFailure(parse_failure))`.